packages/core/lib/commands/test/run.js

```diff
@@ -107,6 +107,27 @@ module.exports = async function (options) {
     }
   };
 
+  function sanitizeGanacheOptions(ganacheOptions) {
+    let network_id = ganacheOptions.network_id;
+
+    // Use default network_id if "*" is defined in config
+    if (network_id === "*") {
+      network_id = 4447;
```

Comment:
Looking at this a bit more -- where did this `4447` come from?  I would think that the thing to do would be, if `network_id` in the config is `"*"`, simply not *put* a network ID in the Ganache options, and let it use its own default, rather than us setting it for it.  Is there a reason for putting 4447 instead?

(Sorry, still not a proper review, but I thought that was worth commenting on!)

Reply:
Thanks @haltman-at for raising this. 4447 is the default network_id we use for ganache in truffle test. When the user specifies network_id in config, it overrides the default ganache options set in the `if` block of the code and then we are calling the `sanitizeGanacheOptions` function. If the network_id is `*` then it again returns the default the ganache network_id, that is 4447.

Reply:
Should we just leave it out in the case "*" is specified? Seems like it might be a bit simpler.

Reply:
It makes sense that `*` should not assign a network_id for ganacheOptions (for test) and allow Ganache to use its default, current timestamp.

Reply:
Since, we are calling the sanitize function after overriding the default ganache options, the network_id will return "*" again if we leave it out.

Reply:
@sukanyaparashar You can *remove* the network ID if `"*"` was passed.  Your options aren't limited to setting a number or letting `"*"` pass through.  You can either overwrite it as `undefined`, or, if that's not sufficient, use the `delete` operator to remove it entirely.

Reply:
What about setting `network_id = Date.now()`  instead of `4777` to align with Ganache's default when `network_id === '*'`?

Reply:
> @sukanyaparashar You can _remove_ the network ID if `"*"` was passed. Your options aren't limited to setting a number or letting `"*"` pass through. You can either overwrite it as `undefined`, or, if that's not sufficient, use the `delete` operator to remove it entirely.

https://github.com/trufflesuite/truffle/blob/2e85aa659d680a6ea7feb5bc7089556b3889557f/packages/environment/environment.js#L132

We have a validation check for network_id == null.

Reply:
> What about passing along `network_id = Date.now()` instead of `4777` to align with Ganache's default when the `networks.test.network_id === '*'`?

That can be done @cds-amal . But should we change it only when `network_id === "*"` or should we set the default ganache `network_id = Date.now()` when even no network is specified in the config?

Reply:
Good question. If network_id = '*' we set it to Date.now(), if network_id is unspecified in the config, then Ganache will set its networkId to Date.now(). I'm ok leaving the special case of only setting it when `config.network_id = '*'` and let Ganache handle the other case. What do others think?

Reply:
I don't think we should use `Date.now()`, because that -- like `4447` or any other attempt on our part to match Ganache's default -- is a fragile solution.  If the network_id is `"*"`, we should let Ganache do all the deciding, not us.  That's the robust solution.  If we try to match, it could change in the future.  (`Date.now()` is particularly worrying because that changes over time...)

@sukanyaparashar I don't see why the check you describe is relevant.  That runs as part of `Environment.detect(config)`.  I'm not saying you should remove the network ID from the config when network_id is `"*"`, but rather that it should be removed from the ganache options.  I don't believe there's any reason we can't do the latter?

Reply:
https://github.com/trufflesuite/truffle/blob/01c1d42d72564bf58b698742bf2ffe4ed0678901/packages/environment/environment.js#L75

I think there is a catch here @haltman-at . If I comment out the line in the `develop` function here, then ganache starts up even without the network_id.

Reply:
I am not suggesting you comment out the line.  I am suggesting that, specifically when `network_id` from the config is equal to `"*"`, you actively remove `network_id` from the ganache options, i.e., by setting it to `undefined` or deleting it.  This requires checking for this condition and acting on it, not just commenting out a line.

Reply:
Oh!  OK the catch has been explained in a meeting -- the problem isn't `Environment.detect`, it's a particular manual copy in `Environment.develop`.  That can be fixed, but I've been convinced that `4447` (or `Date.now()`) is fine.  I'll mark this resolved.